spark_log_parser/extractor.py

```diff
@@ -0,0 +1,213 @@
+                    self._remove_from_stats(tf.size)
+                    extract_path.unlink()
+
+                paths += sub_paths
+
+        return paths
+
+    def _extract_zip(self, archive_path: Path, extract_dir: Path):
+        paths = []
+
+        with zipfile.ZipFile(archive_path) as zfile:
+            for zinfo in zfile.infolist():
+                if self._should_skip_file(zinfo.filename):
+                    continue
+                if zinfo.is_dir():
+                    continue
+
+                self._add_to_stats_and_verify(zinfo.file_size)
+
+                extract_path = extract_dir.joinpath(zinfo.filename)
+                extract_path.parent.mkdir(parents=True, exist_ok=True)
+                with zfile.open(zinfo.filename) as source, open(extract_path, "wb") as target:
+                    shutil.copyfileobj(source, target)
+
+                # Remove the archive if it was extracted to free up space
+                sub_paths = self._extract(extract_path)
+                if extract_path not in sub_paths:
+                    self._remove_from_stats(zinfo.file_size)
+                    extract_path.unlink()
+
+                paths += sub_paths
+
+        return paths
+
+    def _extract_gz(self, archive_path: Path, extract_path: Path):
+        with open(archive_path, "rb") as fobj:
+            fobj.seek(-4, 2)
+            size = struct.unpack("I", fobj.read(4))[0]
+
+        self._add_to_stats_and_verify(size)
+
+        extract_path.parent.mkdir(parents=True, exist_ok=True)
+        with gzip.open(archive_path, "rb") as source, open(extract_path, "wb") as target:
+            shutil.copyfileobj(source, target)
+
+        # Remove the archive if it was extracted to free up space
+        sub_paths = self._extract(extract_path)
+        if extract_path not in sub_paths:
+            self._remove_from_stats(size)
+            extract_path.unlink()
+
+        return sub_paths
+
+    FILE_SKIP_PATTERNS = [".DS_Store".lower(), "__MACOSX".lower(), "/."]
+
+    def _should_skip_file(self, filename: str):
+        if filename.startswith("."):
+            return True
+
+        filename = filename.lower()
+        for name in Extractor.FILE_SKIP_PATTERNS:
+            if name in filename:
+                return True
+
+        return False
+
+    def _add_to_stats_and_verify(self, size, count=1):
+        self.size_total += size
+        self.file_total += count
+
+        ratio = size / self.size_total
+        if ratio > THRESHOLD_RATIO:
+            raise AssertionError("Encountered suspicious compression ratio in the archive")
+
+        if self.size_total > THRESHOLD_SIZE:
+            raise AssertionError("The archive is too big")
+
+        if self.file_total > THRESHOLD_ENTRIES:
+            raise AssertionError("Too many files in the archive")
+
+
+    def _remove_from_stats(self, size, count=1):
+        self.size_total -= size
+        self.file_total -= count
+
+
+    def _download(self, path):
```

Comment:
The only shortcoming I see here is that it requires the s3 to be a single prefix to an archive or single eventlog. It could be expanded to ingesting a prefix, e.g. using the method I implemented here. https://github.com/synccomputingcode/sync_backend/blob/bbb8c286d03b13668b87ff6f68b5aaaace071598/spark_predictor/internal_tools/analysis/file_copy_handler.py#L46

This isn't necessary, IMO, as long as the README makes clear that s3 links must be a full object path and not a prefix.